## JITEN MISHRA DSC430_Assignment0702_OverlapEllipse
### I have not given or received any unauthorized assistance on this assignment.

In [3]:
# install library
!pip install ipynb
import random
import math
import turtle

In [4]:
class Point:
    """
    Class to represent a Point coordinate
    """
    
    def __init__(self, xCor=0.0, yCor=0.0):
        """
        Initialize the x and y coordinate
        """
        
        self.xCor = xCor
        self.yCor = yCor
        
    def getXCor(self):
        """
        Returns the x coordinate
        """
        
        return self.xCor
    
    def getYCor(self):
        """
        Returns the y coordinate
        """
        
        return self.yCor
    
    def __repr__(self):
        """
        String representation of the object
        """
        
        ret = "Point({},{})".format(self.xCor,self.yCor)
        return ret
    

In [14]:
class Ellipse:
    """
    Class to represent an Ellipse 
    """
    
    def __init__(self, pe1, pe2, width=1.0):
        """
        Initialize the x,y coordinates and width of the ellipse
        and the point reference
        """
        
        self.x1 = pe1.getXCor()
        self.x2 = pe2.getXCor()
        self.y1 = pe1.getYCor()
        self.y2 = pe2.getYCor()
        self.width = width
        self.pe1 = pe1
        self.pe2 = pe2
        
    def getx1(self):
        """
        Returns the x1 coordinate
        """
        
        return self.x1
    
    def getx2(self):
        """
        Returns the x2 coordinate
        """
        
        return self.x2
    
    def gety1(self):
        """
        Returns the y1 coordinate
        """
        
        return self.y1
    
    def gety2(self):
        """
        Returns the y2 coordinate
        """
        
        return self.y2
    
    def getWidth(self):
        """
        Returns the width of the ellipse
        """
        
        return self.width
    
    def getArea(self):
        """
        Returns the Area of the ellipse
        """
        
        # the computation of the area of ellipse is written
        # keeping in mind the completion of this assignment
        # generic calculation is more complex when you dont have 
        # the short axis data with you and when ellipse are slanted
        
        # area of a ellipse is π * a * b
        # a is half of long axis
        # b is half of short axis
        a = self.width / 2
        
        # if the the ellipse is on x axis
        # else if it in on y axis
        if(self.x1 == 0 and self.x2 == 0):
            c = abs(self.y2)
        else:
            c = abs(self.x2)

        # formula to compute b =√a^2 - c^2
        b = math.sqrt(abs(a**2 - c**2))
        
        # area of ellipse
        self.area = math.pi * a * b
        
        return self.area
    
    def __repr__(self):
        """
        Return the string representation
        """
        
        ret = "Ellipse({},{},{}) has area {}".format(self.pe1.__repr__(),self.pe2.__repr__(),
                                                     self.width,self.getArea())
        
        return ret
    

In [21]:
# import the specific definition from Assgn 0701
from ipynb.fs.defs.DSC430_Assignment0701_WarAndPeace import WarAndPeacePrng

def computeOverlapOfEllipses(e1, e2):
    """
    Initiates the computation of the area of overlap of two Ellipse
    e1 : 1st Ellipse
    e2 : 2nd Ellipse
    """
    
    # get all the x and y coordinates of focal points
    # and the max width of the ellipse
    xVals,yVals,maxW = getCoordinatesAndMaxWidth(e1, e2)
    
    # compute the world coordinates
    llx,lly,urx,ury = computeBoxCoordinate(xVals, yVals, maxW)
    
    loop = 10000
    
    # commenting out the drawAndCompute call to execute faster
    # this call draws the simulation and computes the output
    # since this is not a requirement for the assignment
    # however this works too
    #overlapArea,countHit,loop = drawAndCompute(llx, lly, urx, ury, e1, e2,loop)
    
    # call the compute function to get the count of hits and overlap area
    overlapArea,countHit = compute(llx, lly, urx, ury, e1, e2, loop)
        
    # print the output in desired format
    printOutput(e1, e2, countHit, overlapArea,loop)
    
    # return the overlap area
    return overlapArea


In [7]:
def getCoordinatesAndMaxWidth(e1, e2):
    """
    Returns the coordinates of focal point of ellipse
    and the maximum width of the ellipse
    e1 : 1st Ellipse Object
    e2 : 2nd Ellipse Object
    """
    
    xVals = []
    yVals = []
    
    # populate the list with all x coordinates
    xVals.append(e1.getx1())
    xVals.append(e1.getx2())
    xVals.append(e2.getx1())
    xVals.append(e2.getx2())
    
    # populate the list with all y coordinates
    yVals.append(e1.gety1())
    yVals.append(e1.gety2())
    yVals.append(e2.gety1())
    yVals.append(e2.gety2())
    
    # get the width of the ellipse
    w1 = e1.getWidth()
    w2 = e2.getWidth()
    
    maxW = max(w1,w2)
        
    # return the lists of coordinates
    return xVals,yVals,maxW
    

In [8]:
def computeBoxCoordinate(xVals, yVals, maxW):
    """
    Compute the world coordinate to plot a box
    xVals : list of all x coordinates
    yVals : list of all y coordinates
    maxW  : maximum width of the ellipse
    """
    
    # compute the worldcoordinates of box
    # llx: a number, x-coordinate of lower left corner of canvas
    # lly: a number, y-coordinate of lower left corner of canvas
    # urx: a number, x-coordinate of upper right corner of canvas
    # ury: a number, y-coordinate of upper right corner of canvas
    llx = min(xVals) - (maxW/2)
    lly = min(yVals) - (maxW/2)
    urx = max(xVals) + (maxW/2)
    ury = max(yVals) + (maxW/2)
    
    
    return llx,lly,urx,ury

In [9]:
def drawAndCompute(llx, lly, urx, ury, e1, e2, loop):
    """
    Function to draw a visual plot of the random points
    And compute the overlap of area of two ellipse
    llx  : lower left x of complete area
    lly  : lower left y of complete area
    urx  : uper right x of complte area
    urx  : uper right y of complte area
    e1   : 1st Ellipse
    e2   : 2nd Ellipse
    loop : total iteration count
    """
    
    # compute the scalling length
    scaleLength = urx - llx
    
    # compute the scalling height
    scaleHeight = ury - lly
    
    wn = turtle.Screen()
    drawingT = turtle.Turtle()
    
    wn.setworldcoordinates(llx,lly, urx, ury)
    
    # get random points
    wnp = WarAndPeacePrng()
    
    countHit = 0
    drawingT.up()
    
    for i in range(loop):
        
        # get a random point x,y and
        # scale the x,y to match the box
        x = (wnp.random() * scaleLength) + llx 
        y = (wnp.random() * scaleHeight) + lly
        
        # initialize the poit object
        randPoint = Point(x,y)
        
        # check if the random point is inside the ellipses
        e1Flag = isInsideEllipse(e1,randPoint)
        e2Flag = isInsideEllipse(e2,randPoint)
        
        # shift the turtle to the point
        drawingT.goto(x,y)
                  
        # compute the colors for the dot as per the point location
        # red for overlap , blue for ellipse 1 and green for ellipse 2
        # else black
        if ( e1Flag and e2Flag ):
            drawingT.color("red")
            countHit += 1
        elif(e1Flag):
            drawingT.color("blue")
        elif(e2Flag):
            drawingT.color("green")
        else:
            drawingT.color("black")
            
        # print the dot
        drawingT.dot()
    
    wnp.close()
    
    # calculate the overlap Area
    overlapArea = countHit/loop*scaleLength*scaleHeight
    
    # close the window
    wn.exitonclick()
        
    # return overlap Area and count of hits
    return overlapArea,countHit

In [10]:
def compute(llx, lly, urx, ury, e1, e2, loop):
    """
    Function to compute the overlap of area of two ellipse
    llx  : lower left x of complete area
    lly  : lower left y of complete area
    urx  : uper right x of complte area
    urx  : uper right y of complte area
    e1   : 1st Ellipse
    e2   : 2nd Ellipse
    loop : total iteration count
    """
    
     # compute the scalling length
    scaleLength = urx - llx
    
    # compute the scalling height
    scaleHeight = ury - lly
    
    # initialize random points object
    wnp = WarAndPeacePrng()
    
    countHit = 0
    
    for i in range(loop):
        
        # get a random point x,y and
        # scale the x,y to match the box
        x = (wnp.random() * scaleLength) + llx 
        y = (wnp.random() * scaleHeight) + lly
        
        # initialize the poit object
        randPoint = Point(x,y)
        
        # check if the random point is inside ellipse 1
        e1Flag = isInsideEllipse(e1,randPoint)
        
        # check if the random point is inside ellipse 2
        e2Flag = isInsideEllipse(e2,randPoint)
        
        # if the point is in both the ellipse
        # increase the hit count
        if(e1Flag and e2Flag):
            countHit += 1
        
    wnp.close()
    
    # calculate the overlap area
    overlapArea = countHit/loop*scaleLength*scaleHeight
    
    # return the area,hit count and total loop count
    return overlapArea,countHit


In [11]:
def isInsideEllipse(e, p):
    """
    Return True or False for if the point is inside Ellipse or not
    e : Ellipse object
    p : Point object
    """
    
    # compute the distance of the point 
    # from both focal point of the ellipse
    # distance of two points = √((x2-x1)^2 + (y1-y2)^2)
    focal1ToPdist = math.sqrt((e.getx1() - p.getXCor())**2 + (e.gety1() - p.getYCor())**2 )
    focal2ToPdist = math.sqrt((e.getx2() - p.getXCor())**2 + (e.gety2() - p.getYCor())**2 )
    
    # get the width of the ellipse
    eWidth = e.getWidth()
    
    # if the sum of distance of the point is < width
    # the the point is inside the ellipse
    if(focal1ToPdist + focal2ToPdist < eWidth):
        return True
    else:
        return False
    

In [12]:
def printOutput(e1, e2, countHit, overlapArea, loop):
    """
    Print the Output of the test
    """
    
    print("***************************************************************************")
    print(e1)
    print(e2)
    print("{} out of {} generated points are in both ellipses".format(countHit, loop))
    print("The overlap of the two has area {}".format(overlapArea))
    print("***************************************************************************")
    

In [22]:
# Test 1 : two circles at the origin
p = Point(0,0)
p1 = Point(0,0)

p2 = Point(0,0)
p3 = Point(0,0)

e1 = Ellipse(p,p1,2)
e2 = Ellipse(p2,p3,4)

overlap = computeOverlapOfEllipses(e1,e2)


***************************************************************************
Ellipse(Point(0,0),Point(0,0),2) has area 3.141592653589793
Ellipse(Point(0,0),Point(0,0),4) has area 12.566370614359172
1942 out of 10000 generated points are in both ellipses
The overlap of the two has area 3.1072
***************************************************************************


#### Varification of Result
* The Ellipse are circle since the focal points meet
* The 1st Ellipse(circle) has diameter 2 so the area should be π\*r^2 i.e (3.1415) and we can see that in the result
* The 2nd Ellipse(circle) has diameter 4 so the area should be π\*r^2 i.e (12.566) and we can see that in the result
* We tried to dart 10000 points on the box out of which 1953 points were those points that fell in both ellipse 1 and 2
* And we got the area of overlap as 3.1072 which is almost the area of inner circle 3.1415 
* If we increase the dart points then the value will match exactly as 3.1415

In [18]:
# Test 2 : One circle and One ellipse with same y axis
# here the ellipse is inside the circle
p = Point(0,0)
p1 = Point(0,0)

p2 = Point(-1,0)
p3 = Point(1,0)

e1 = Ellipse(p,p1,4)
e2 = Ellipse(p2,p3,4)

overlap = computeOverlapOfEllipses(e1,e2)

***************************************************************************
Ellipse(Point(0,0),Point(0,0),4) has area 12.566370614359172
Ellipse(Point(-1,0),Point(1,0),4) has area 10.882796185405306
4558 out of 10000 generated points are in both ellipses
The overlap of the two has area 10.9392
***************************************************************************


#### Varification of Result
* Both the ellipse has both the y axis common 
* The 1st Ellipse(circle) has diameter 4 so the area should be π*r^2 i.e (12.566) and we can see that in the result
* The 2nd Ellipse has diameter 4 so the area should be π\*a\*b i.e (10.882) and we can see that in the result
* Here the 2nd Ellipse falls inside the 1st one hence the common area is the area of 2nd Ellipse
* We tried to dart 10000 points on the box out of which 4514 points were those points that fell in both ellipse 1 and 2
* And we got the area of overlap as 10.9392 which is almost the area of 2nd Ellipse 10.8827 

In [43]:
# Test 3 : Two ellipse perpendicular to each other on the y axis
# here part of the ellipse 2 is inside ellipse 1
# This test was ran 100000 by edditing the loop count in the code to get a drawing figure
# and the figure is attached in the submission. however the code is commented here w.r.t drawing 
p = Point(-1,0)
p1 = Point(1,0)

p2 = Point(0,0)
p3 = Point(0,-3)

e1 = Ellipse(p,p1,4)
e2 = Ellipse(p2,p3,5)

overlap = computeOverlapOfEllipses(e1,e2)

***************************************************************************
Ellipse(Point(-1,0),Point(1,0),4) has area 10.882796185405306
Ellipse(Point(0,0),Point(0,-3),5) has area 13.02435509511789
1342 out of 10000 generated points are in both ellipses
The overlap of the two has area 7.515200000000001
***************************************************************************


#### Varification of Result
* Both the ellipse are perpendicular to each other haaving some common area oraound the origin. 
* The 1st Ellipse has diameter 4 so the area should be π\*a\*b i.e (10.882) and we can see that in the result
* The 2nd Ellipse has diameter 5 so the area should be π\*a\*b i.e (13.024) and we can see that in the result
* Here only a part of the 2nd Ellipse falls inside the 1st one hence the common area is always less than any of the area of ellipse
* We tried to dart 10000 points on the box out of which 1342 points were those points that fell in both ellipse 1 and 2
* And we got the area of overlap as 7.5152
* If we increase the dart points also still its never going to match the area of any of the ellipse.

In [44]:
# Test 4 : Two Ellipse away from each other 
# No overlapping scenario
p = Point(0,0)
p1 = Point(0,-3)

p2 = Point(-1,10)
p3 = Point(2,10)

e1 = Ellipse(p,p1,5)
e2 = Ellipse(p2,p3,6)

overlap = computeOverlapOfEllipses(e1,e2)

***************************************************************************
Ellipse(Point(0,0),Point(0,-3),5) has area 13.02435509511789
Ellipse(Point(-1,10),Point(2,10),6) has area 21.07444419312218
0 out of 10000 generated points are in both ellipses
The overlap of the two has area 0.0
***************************************************************************


#### Varification of Result
* Both the ellipse are way away from each other. 
* The 1st Ellipse has diameter 5 so the area should be π\*a\*b i.e (13.024) and we can see that in the result
* The 2nd Ellipse has diameter 5 so the area should be π\*a\*b i.e (21.074) and we can see that in the result
* We tried to dart 10000 points on the box out of which 0 points were those points that fell in both ellipse 1 and 2
* And we got the area of overlap as 0
* Since the ellipse are away from each other there are no overlapping and we see that in the result.